declarations_site/catalog/utils.py

```diff
@@ -25,6 +26,103 @@ def replace_apostrophes(s):
+
+
+def apply_search_sorting(search, sort=""):
+    sort_keys = {'name': 'general.full_name.raw', 'year': 'intro.declaration_year'}
+    if sort and sort in sort_keys:
+        search = search.sort(sort_keys[sort])
+    return search
+
+
+def apply_term_filter(search, filter_value, field):
+    if filter_value:
+        filter_kw = {field: filter_value}
+        search = search.filter("term", **filter_kw)
+    return search
+
+
+def apply_match_filter(search, filter_list, field, operator='or'):
+    if filter_list:
+        value = " ".join(filter_list)
+        filter_kw = {field: {'query': value, 'operator': operator}}
+        search = search.filter("match", **filter_kw)
+    return search
+
+
+def apply_search_filters(search, filters):
+    region_type = filters.get("region_type", "")
+    region_value = filters.get("region_value", "")
+    if region_type and region_value:
+        filters[region_type] = region_value
+    search = apply_term_filter(search, filters.get("declaration_year", ""), "intro.declaration_year")
+    search = apply_term_filter(search, filters.get("doc_type", ""), "intro.doc_type")
+    search = apply_term_filter(search, filters.get("region", ""), "general.post.region.raw")
+    search = apply_term_filter(search, filters.get("actual_region"), "general.post.actual_region.raw")
+    search = apply_term_filter(search, filters.get("estate_region"), "estate.region.raw")
+    search = apply_match_filter(search, filters.getlist("post_type"), "general.post.post_type")
+    return search
+
+
+QS_OPS = re.compile(r'(["*?:~(]| AND | OR | NOT | -\w)')
+QS_NOT = re.compile(r'[/]')  # do not allow regex queries
+
+
+def base_search_query(base_search, query, deepsearch, filters):
+    if not query:
+        search = base_search.query('match_all')
+        search = apply_search_filters(search, filters)
+        return search
+
+    default_field = "_all" if deepsearch else "index_card"
+
+    # try Lucene syntax query first
+    if QS_OPS.search(query) and not QS_NOT.search(query):
+        query_kw = {
+            "analyze_wildcard": True,
+            "allow_leading_wildcard": False,
+            "default_field": default_field,
+            "default_operator": "and",
+            "query": query
+        }
+
+        search = base_search.query("query_string", **query_kw)
+        search = apply_search_filters(search, filters)
+
+        try:
+            if search.count():
+                return search
+        except TransportError:
+            # if Lucene query fail return to regular
+            pass
+
+        # before pass to match_query remove all unnecessary chars
+        query = re.sub(r'["*?:~()[\]/]', '', query)
+
+    query_op = {"query": query, "operator": "and"}
+    query_kw = {default_field: query_op}
+
+    search = base_search.query("match", **query_kw)
+    search = apply_search_filters(search, filters)
+
+    nwords = len(re.findall(r'\w{4,}', query))
+
+    if nwords > 2 and deepsearch and not search.count():
```

Comment:
1. apply only if 2 or more words in query
2. minimum_should_match = num_words - 1  # or - 2 for 5 or more words